### PR TITLE
add ngc image signing job for auto signing

### PR DESCRIPTION
**Pull request by shivakunv**
Auto siging of gpu-operator released images @cdesiniotis

### CODE REVIEW

**Comment by shivakunv** [comment]
@cdesiniotis PTAL

**Comment by cdesiniotis** [comment]
Thank you @shivakunv this lgtm. 

@elezar these changes will automatically sign our container images in our NGC release pipelines. Shiva was able to verify the pipeline logic in this pipeline: https://gitlab-master.nvidia.com/dl/container-dev/container-toolkit/-/pipelines/15401703

**Comment by cdesiniotis** [comment]
@shivakunv can you please sign your commit.

File: .nvidia-ci.yml [hunk 1]
```diff
@@ -244,3 +244,69 @@ release:ngc-packaging:
...
+  stage: sign
+  variables:
+    IMAGE_NAME: "${OUT_IMAGE_NAME}"
+    IMAGE_TAG: "${OUT_IMAGE_TAG}"
+  retry:
+    max: 2
+  before_script:
+    - !reference [.ngccli-setup, before_script]
+    # We ensure that the OUT_IMAGE_TAG is set
+    - 'echo Image Tag: ${IMAGE_NAME} ; [[ -n "${IMAGE_TAG}" ]] || exit 1'
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:273]:
```suggestion
    - 'echo Image Tag: ${IMAGE_TAG} ; [[ -n "${IMAGE_TAG}" ]] || exit 1'
```

**Comment by shivakunv** [comment]
signed and corrected the echo command

**Comment by elezar** [comment]
@shivakunv I think @cdesiniotis meant "sign-off". Since we use the DCO for contributions, you need to sign-off your commits using `git commit -s` (in addition to or instead of `-S`).

File: .nvidia-ci.yml [hunk 2]
```diff
@@ -244,3 +244,69 @@ release:ngc-packaging:
...
+    - curl -sSLo ngccli_linux.zip https://api.ngc.nvidia.com/v2/resources/nvidia/ngc-apps/ngc_cli/versions/${NGCLI_VERSION}/files/ngccli_linux.zip
+    - unzip ngccli_linux.zip
+    - chmod u+x ngc-cli/ngc
+    - export PATH=$(pwd)/ngc-cli:${PATH}
+    - NGC_CLI_API_KEY=${NGC_REGISTRY_TOKEN} ngc version upgrade --org nvidia -y
+
+# .sign forms the base of the deployment jobs which signs images in the CI registry.
+# This is extended with the image name and version to be deployed.
+.sign:ngc:
+  image: ubuntu:latest
```
Comment (elezar) [inline .nvidia-ci.yml:263]: Question: Does the `ngccli` need Ubuntu, or would the busybox-based image that we use for our other steps be sufficient?

File: .nvidia-ci.yml [hunk 3]
```diff
@@ -244,3 +244,69 @@ release:ngc-packaging:
...
+# Download the ngc cli binary for use in the sign steps
+.ngccli-setup:
+  before_script:
+    - export NGCLI_VERSION=3.41.4
+    - apt-get update && apt-get install -y curl unzip jq
+    - curl -sSLo ngccli_linux.zip https://api.ngc.nvidia.com/v2/resources/nvidia/ngc-apps/ngc_cli/versions/${NGCLI_VERSION}/files/ngccli_linux.zip
+    - unzip ngccli_linux.zip
+    - chmod u+x ngc-cli/ngc
+    - export PATH=$(pwd)/ngc-cli:${PATH}
+    - NGC_CLI_API_KEY=${NGC_REGISTRY_TOKEN} ngc version upgrade --org nvidia -y
```
Comment (elezar) [inline .nvidia-ci.yml:258]: Question: What does `ngc version upgrade` do?

File: .nvidia-ci.yml [hunk 4]
```diff
@@ -244,3 +244,69 @@ release:ngc-packaging:
...
+  extends:
+    - .dist-ubuntu20.04
+    - .sign:ngc
+  needs:
+    - release:ngc-ubuntu20.04
+  rules:
+    - if: $CI_COMMIT_TAG
+  variables:
+    OUT_IMAGE_TAG: "${CI_COMMIT_TAG}-${DIST}"
+    OUT_IMAGE_NAME: "${NGC_REGISTRY_IMAGE}"
```
Comment (elezar) [inline .nvidia-ci.yml:288]: The `rules` and `variables` can be specified ONCE in `.sign:ngc` instead of for each job.

File: .nvidia-ci.yml [hunk 5]
```diff
@@ -244,3 +244,69 @@ release:ngc-packaging:
   extends:
     - .dist-packaging
     - .release:ngc
+
+# Define the external image signing steps for NGC
+# Download the ngc cli binary for use in the sign steps
+.ngccli-setup:
+  before_script:
+    - export NGCLI_VERSION=3.41.4
+    - apt-get update && apt-get install -y curl unzip jq
```
Comment (elezar) [inline .nvidia-ci.yml:253]: Question: why is `jq` required?

**Review by elezar — COMMENTED** [review]
Thanks @shivakunv.

Some comments / questions.

File: .nvidia-ci.yml (hunk 2 above)
Comment (shivakunv) [reply .nvidia-ci.yml:263]: ubuntu needed , busybox version will give undefined reference 
* ngc dependency issue due to busybox image . ngc built with latest toolchain and busybox is on older version.
* `ldd $(pwd)/ngc-cli/ngc /lib64/ld-linux-x86-64.so.2 (0x7f55b916c000) libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x7f55b916c000) libz.so.1 => /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/libz.so.1 (0x7f55b8f56000) libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x7f55b916c000) libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x7f55b916c000) Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/libz.so.1: __vsnprintf_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/libz.so.1: __sprintf_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __strcat_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __snprintf_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __vfprintf_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __realpath_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __strdup: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __memcpy_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __vsnprintf_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __strcpy_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __fread_chk: symbol not found Error relocating /builds/JQq79Cy1/0/dl/container-dev/gpu-operator/ngc-cli/ngc: __fprintf_chk: symbol not found`

File: .nvidia-ci.yml (hunk 5 above)
Comment (shivakunv) [reply .nvidia-ci.yml:253]: not required , I'll remove it

File: .nvidia-ci.yml (hunk 4 above)
Comment (shivakunv) [reply .nvidia-ci.yml]: done

File: .nvidia-ci.yml (hunk 5 above)
Comment (shivakunv) [reply .nvidia-ci.yml]: removed

**Comment by shivakunv** [comment]
@elezar 
PTAL : I updated the code and comment

**Comment by shivakunv** [comment]
> @shivakunv I think @cdesiniotis meant "sign-off". Since we use the DCO for contributions, you need to sign-off your commits using `git commit -s` (in addition to or instead of `-S`).

done

File: .nvidia-ci.yml (hunk 3 above)
Comment (elezar) [reply .nvidia-ci.yml:258]: Is there a way to determine the latest version through the API? Is `https://api.ngc.nvidia.com/v2/resources/nvidia/ngc-apps/ngc_cli/versions/latest/files/ngccli_linux.zip` valid, for example?

File: .nvidia-ci.yml [hunk 6]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+    IMAGE_NAME: "${NGC_REGISTRY_IMAGE}"
+    IMAGE_TAG: "${CI_COMMIT_TAG}"
+  retry:
+    max: 2
+  before_script:
+    - !reference [.ngccli-setup, before_script]
+    - |
+      if [ -n "$DIST" ]; then
+        export IMAGE_TAG="${IMAGE_TAG}-${DIST}"
+      fi
```
Comment (elezar) [inline .nvidia-ci.yml:277]: When is this condition ever true? We seem to always use this where `DIST` is set.

Let's rather remove this and specify `IMAGE_TAG` as `${CI_COMMIT_TAG}-${DIST}` for other cases where we have the "short" tag, we can override ONLY `IMAGE_TAG` in the job that extends this one.

File: .nvidia-ci.yml [hunk 7]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+# Define the external image signing steps for NGC
+# Download the ngc cli binary for use in the sign steps
+.ngccli-setup:
+  before_script:
+    - export NGCLI_VERSION=3.41.4
+    - apt-get update && apt-get install -y curl unzip
+    - curl -sSLo ngccli_linux.zip https://api.ngc.nvidia.com/v2/resources/nvidia/ngc-apps/ngc_cli/versions/${NGCLI_VERSION}/files/ngccli_linux.zip
+    - unzip ngccli_linux.zip
+    - chmod u+x ngc-cli/ngc
+    - export PATH=$(pwd)/ngc-cli:${PATH}
```
Comment (elezar) [inline .nvidia-ci.yml:257]: Is setting the path required? Does this work as expected if we just specify the absolute path as `ngc-cli/ngc` when running the `ngc` command?

File: .nvidia-ci.yml [hunk 8]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+    - NGC_CLI_API_KEY=${NGC_REGISTRY_TOKEN} ngc version upgrade --org nvidia -y
```
Comment (elezar) [inline .nvidia-ci.yml:258]: Question: Should we rather set this as a variable to reduce the risk of leaking this on the command line?

File: .nvidia-ci.yml [hunk 9]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+# .sign forms the base of the deployment jobs which signs images in the CI registry.
+# This is extended with the image name and version to be deployed.
+.sign:ngc:
+  image: ubuntu:latest
+  stage: sign
+  rules:
+    - if: $CI_COMMIT_TAG
+  variables:
+    IMAGE_NAME: "${NGC_REGISTRY_IMAGE}"
+    IMAGE_TAG: "${CI_COMMIT_TAG}"
```
Comment (elezar) [inline .nvidia-ci.yml:269]:
```suggestion
  variables:
    IMAGE_NAME: "${NGC_REGISTRY_IMAGE}"
    IMAGE_TAG: "${CI_COMMIT_TAG}-${DIST}"
    NGC_CLI_API_KEY: "${NGC_REGISTRY_TOKEN}"
```

File: .nvidia-ci.yml [hunk 10]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+  retry:
+    max: 2
+  before_script:
+    - !reference [.ngccli-setup, before_script]
+    - |
+      if [ -n "$DIST" ]; then
+        export IMAGE_TAG="${IMAGE_TAG}-${DIST}"
+      fi
+    # We ensure that the OUT_IMAGE_TAG is set
+    - 'echo Image Tag: ${IMAGE_TAG} ; [[ -n "${IMAGE_TAG}" ]] || exit 1'
```
Comment (elezar) [inline .nvidia-ci.yml:279]: Should we do the same for `IMAGE_NAME`?

File: .nvidia-ci.yml [hunk 11]
```diff
@@ -244,3 +244,60 @@ release:ngc-packaging:
...
+    IMAGE_TAG: "${CI_COMMIT_TAG}"
+  retry:
+    max: 2
+  before_script:
+    - !reference [.ngccli-setup, before_script]
+    - |
+      if [ -n "$DIST" ]; then
+        export IMAGE_TAG="${IMAGE_TAG}-${DIST}"
+      fi
+    # We ensure that the OUT_IMAGE_TAG is set
```
Comment (elezar) [inline .nvidia-ci.yml:278]: This comment is inaccurate. We are checking a different variable.

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks for the update @shivakunv.

Some additional comments. Also, please squash your commits or sign-off on all of them.

File: .nvidia-ci.yml (hunk 5 above)
Comment (shivakunv) [reply .nvidia-ci.yml:252]: to get latest version of ngc

**Pull request merged**